lib/chef/knife/bootstrap/client_builder.rb

```diff
@@ -0,0 +1,190 @@
+        end
+
+        private
+
+        # @return [String] node name from the knife_config
+        def node_name
+          knife_config[:chef_node_name]
+        end
+
+        # @return [String] enviroment from the knife_config
+        def environment
+          knife_config[:environment]
+        end
+
+        # @return [String] run_list from the knife_config
+        def run_list
+          knife_config[:run_list]
+        end
+
+        # @return [Hash,Array] Object representation of json first-boot attributes from the knife_config
+        def first_boot_attributes
+          knife_config[:first_boot_attributes]
+        end
+
+        # @return [String] chef server url from the Chef::Config
+        def chef_server_url
+          chef_config[:chef_server_url]
+        end
+
+        # Accesses the run_list and coerces it into an Array, changing nils into
+        # the empty Array, and splitting strings representations of run_lists into
+        # Arrays.
+        #
+        # @return [Array] run_list coerced into an array
+        def normalized_run_list
+          case run_list
+          when nil
+            []
+          when String
+            run_list.split(/\s*,\s*/)
+          when Array
+            run_list
+          end
+        end
+
+        # Create the client object and save it to the Chef API
+        def create_client!
+          Chef::ApiClient::Registration.new(node_name, client_path, http_api: rest).run
+        end
+
+        # Create the node object (via the lazy accessor) and save it to the Chef API
+        def create_node!
+          node.save
+        end
+
+        # Create a new Chef::Node.  Supports creating the node with its name, run_list, attributes
+        # and environment.  This injects a rest object into the Chef::Node which uses the client key
+        # for authentication so that the client creates the node and therefore we get the acls setup
+        # correctly.
+        #
+        # @return [Chef::Node] new chef node to create
+        def node
+          @node ||=
+            begin
+              node = Chef::Node.new(chef_server_rest: client_rest)
+              node.name(node_name)
+              node.run_list(normalized_run_list)
+              node.normal_attrs = first_boot_attributes if first_boot_attributes
+              node.environment(environment) if environment
+              node
+            end
+        end
+
+        # Check for the existence of a node and/or client already on the server.  If the node
+        # already exists, we must delete it in order to proceed so that we can create a new node
+        # object with the permissions of the new client.  There is a use case for creating a new
+        # client and wiring it up to a precreated node object, but we do currently support that.
+        #
+        # We prompt the user about what to do and will fail hard if we do not get confirmation to
+        # delete any prior node/client objects.
+        def sanity_check
```

Comment:
I don't think it should block this for now, but perhaps our "model" classes should just provide this functionality via class methods.  `Chef::Client.exists?(name)`, `Chef::Client.delete(name)`, `Chef::Client.delete_f(name)` (or something for "delete but I don't care if it isn't there").